<a href="https://colab.research.google.com/github/ParkStrat/Sales_Predictions/blob/main/Part2_Project1_Sales_Predictions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [93]:
import pandas as pd

filename = '/content/drive/MyDrive/Coding Dojo - Data Science/01. Week 1: Python/sales_predictions.csv'
df = pd.read_csv(filename)
df.head()

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
0           FDA15         9.30  ...  Supermarket Type1          3735.1380
1           DRC01         5.92  ...  Supermarket Type2           443.4228
2           FDN15        17.50  ...  Supermarket Type1          2097.2700
3           FDX07        19.20  ...      Grocery Store           732.3800
4           NCD19         8.93  ...  Supermarket Type1           994.7052

[5 rows x 12 columns]

# Data Cleaning

In [94]:
# Using .info() to do initial assessment of the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


Dataframe contains
*   Columns: 12
*   Rows: 8523

The below columns are likely missing data:
*   Item_Weight (1,460 values)
*   Outlet_Size (2,410 values)


Columns

0    Item_Identifier            8523 non-null   object

 1   Item_Weight                7060 non-null   float64
 
 2   Item_Fat_Content           8523 non-null   object 
 
 3   Item_Visibility            8523 non-null   float64
 
 4   Item_Type                  8523 non-null   object 
 
 5   Item_MRP                   8523 non-null   float64
 
 6   Outlet_Identifier          8523 non-null   object 
 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 
 8   Outlet_Size                6113 non-null   object 
 
 9   Outlet_Location_Type       8523 non-null   object 
 
 10  Outlet_Type                8523 non-null   object 
 
 11  Item_Outlet_Sales          8523 non-null   float64





In [95]:
# Checking to see if there are any duplicates in the data set
df.duplicated().any()

False

Dataframe contains no duplicates

In [96]:
# Assessing other data associated with missing Outlet Size information
missingSize_filter = df.loc[df['Outlet_Size'].isna()]
missingSize_filter.head()

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
3            FDX07       19.200  ...      Grocery Store           732.3800
8            FDH17       16.200  ...  Supermarket Type1          1076.5986
9            FDU28       19.200  ...  Supermarket Type1          4710.5350
25           NCD06       13.000  ...  Supermarket Type1           838.9080
28           FDE51        5.925  ...      Grocery Store           178.4344

[5 rows x 12 columns]

In [97]:
missingSize_filter['Outlet_Identifier'].value_counts()

OUT045    929
OUT017    926
OUT010    555
Name: Outlet_Identifier, dtype: int64

In [98]:
df['Outlet_Identifier'].value_counts()

OUT027    935
OUT013    932
OUT035    930
OUT049    930
OUT046    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    528
Name: Outlet_Identifier, dtype: int64

The below Outlets have no indication of outlet size:

*   OUT010 (missing 555 of 555 entries)
*   OUT017 (missing 926 of 926 entries)
*   OUT045 (missing 930 of 930 entires)





In [99]:
missingWeight_slice = df.loc[df['Item_Weight'].isna()]
missingWeight_slice.head()

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
7            FDP10          NaN  ...  Supermarket Type3          4022.7636
18           DRI11          NaN  ...  Supermarket Type3          2303.6680
21           FDW12          NaN  ...  Supermarket Type3          4064.0432
23           FDC37          NaN  ...      Grocery Store           214.3876
29           FDC14          NaN  ...      Grocery Store           125.8362

[5 rows x 12 columns]

In [100]:
missingWeight_slice['Outlet_Identifier'].value_counts()

OUT027    935
OUT019    528
Name: Outlet_Identifier, dtype: int64

In [101]:
df.loc[5:22, 'Item_Weight']

5     10.395
6     13.650
7        NaN
8     16.200
9     19.200
10    11.800
11    18.500
12    15.100
13    17.600
14    16.350
15     9.000
16    11.800
17     9.000
18       NaN
19    13.350
20    18.850
21       NaN
22    14.600
Name: Item_Weight, dtype: float64

The below Outlets have no items weights :

*   OUT019 (missing 528 of 528 entries)
*   OUT027 (missing 935 of 935 entries)

Since item properties will likely be critical in the analysis I will be doing for the retailer and since the surrounding data cannot be used to extrapolate a reasonable weight for the missing data, I will remove all the rows for Outlets OUT019 and OUT027 to ensure that only complete data on item specification exists in the dataframe  

In [102]:
df.shape

(8523, 12)

In [103]:
df_minus_Weight = df.loc[~df['Item_Weight'].isna()]
df_minus_Weight.shape

(7060, 12)

In [104]:
df_minus_Weight['Item_Weight'].isna().sum()

0

All the rows without Item_Weight, removed from the dataframe

In [105]:
grocery_filter = df['Outlet_Type'] == 'Grocery Store' 

In [106]:
grocery_slice = df[grocery_filter]
grocery_slice.head()

Item_Identifier  Item_Weight  ...    Outlet_Type  Item_Outlet_Sales
3            FDX07       19.200  ...  Grocery Store           732.3800
23           FDC37          NaN  ...  Grocery Store           214.3876
28           FDE51        5.925  ...  Grocery Store           178.4344
29           FDC14          NaN  ...  Grocery Store           125.8362
30           FDV38       19.250  ...  Grocery Store           163.7868

[5 rows x 12 columns]

In [107]:
grocery_slice['Outlet_Identifier'].value_counts()

OUT010    555
OUT019    528
Name: Outlet_Identifier, dtype: int64

Only OUT010 and OUT019 are grocery stores. There is no way other than requesting more data to determine the store size for OUT010, OUT019, and OUT045.

In [108]:
df_minus_size = df.loc[~df['Outlet_Size'].isna()]
df_minus_size.head()

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
0           FDA15        9.300  ...  Supermarket Type1          3735.1380
1           DRC01        5.920  ...  Supermarket Type2           443.4228
2           FDN15       17.500  ...  Supermarket Type1          2097.2700
4           NCD19        8.930  ...  Supermarket Type1           994.7052
5           FDP36       10.395  ...  Supermarket Type2           556.6088

[5 rows x 12 columns]

In [109]:
# Filter out missing item weights and store sizes
df = df_minus_Weight.loc[~df_minus_Weight['Outlet_Size'].isna()]
df.head()

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
0           FDA15        9.300  ...  Supermarket Type1          3735.1380
1           DRC01        5.920  ...  Supermarket Type2           443.4228
2           FDN15       17.500  ...  Supermarket Type1          2097.2700
4           NCD19        8.930  ...  Supermarket Type1           994.7052
5           FDP36       10.395  ...  Supermarket Type2           556.6088

[5 rows x 12 columns]

In [110]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4650 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            4650 non-null   object 
 1   Item_Weight                4650 non-null   float64
 2   Item_Fat_Content           4650 non-null   object 
 3   Item_Visibility            4650 non-null   float64
 4   Item_Type                  4650 non-null   object 
 5   Item_MRP                   4650 non-null   float64
 6   Outlet_Identifier          4650 non-null   object 
 7   Outlet_Establishment_Year  4650 non-null   int64  
 8   Outlet_Size                4650 non-null   object 
 9   Outlet_Location_Type       4650 non-null   object 
 10  Outlet_Type                4650 non-null   object 
 11  Item_Outlet_Sales          4650 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 472.3+ KB


All rows with missing data removed

In [111]:
df['Item_Fat_Content'].value_counts()

Low Fat    2774
Regular    1575
LF          177
reg          71
low fat      53
Name: Item_Fat_Content, dtype: int64

In [112]:
# Correcting spelling / abbreviation errors for Item_Fat_Content
df.replace({'LF': 'Low Fat', 'low fat': 'Low Fat', 'reg': 'Regular'}, inplace = True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4389: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  method=method,


In [113]:
# Confirming correction
df['Item_Fat_Content'].value_counts()

Low Fat    3004
Regular    1646
Name: Item_Fat_Content, dtype: int64

In [114]:
df['Item_Type'].value_counts()

Fruits and Vegetables    670
Snack Foods              656
Household                498
Frozen Foods             477
Dairy                    380
Canned                   361
Baking Goods             351
Health and Hygiene       277
Soft Drinks              241
Meat                     221
Breads                   132
Hard Drinks              117
Others                    91
Starchy Foods             87
Breakfast                 58
Seafood                   33
Name: Item_Type, dtype: int64

In [115]:
df['Outlet_Location_Type'].value_counts()

Tier 1    1860
Tier 3    1860
Tier 2     930
Name: Outlet_Location_Type, dtype: int64

In [116]:
df.head()

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
0           FDA15        9.300  ...  Supermarket Type1          3735.1380
1           DRC01        5.920  ...  Supermarket Type2           443.4228
2           FDN15       17.500  ...  Supermarket Type1          2097.2700
4           NCD19        8.930  ...  Supermarket Type1           994.7052
5           FDP36       10.395  ...  Supermarket Type2           556.6088

[5 rows x 12 columns]

All categories of data are now consistant

The below sections calculate the min, max, and mean of a numerical columns in the dataframe

In [117]:
df['Item_Weight'].describe()

count    4650.000000
mean       12.898675
std         4.670973
min         4.555000
25%         8.770000
50%        12.650000
75%        17.000000
max        21.350000
Name: Item_Weight, dtype: float64

In [118]:
df['Item_Visibility'].describe()

count    4650.000000
mean        0.060700
std         0.044607
min         0.000000
25%         0.025968
50%         0.049655
75%         0.088736
max         0.188323
Name: Item_Visibility, dtype: float64

In [119]:
df['Item_MRP'].describe()

count    4650.000000
mean      141.716328
std        62.420534
min        31.490000
25%        94.409400
50%       142.979900
75%       186.614150
max       266.888400
Name: Item_MRP, dtype: float64

In [120]:
df['Outlet_Establishment_Year'].describe()

count    4650.000000
mean     1999.190538
std         7.388800
min      1987.000000
25%      1997.000000
50%      1999.000000
75%      2004.000000
max      2009.000000
Name: Outlet_Establishment_Year, dtype: float64

In [121]:
df['Item_Outlet_Sales'].describe()

count     4650.000000
mean      2272.037489
std       1497.964740
min         69.243200
25%       1125.202000
50%       1939.808300
75%       3111.616300
max      10256.649000
Name: Item_Outlet_Sales, dtype: float64